archive/demos/dj.js

```diff
@@ -30,229 +31,263 @@ async function fetchAndDecodeAudio(url) {
-  newSource.buffer = buffer;
+  start(time) {
+    this.source.start(time);
+    this.playing = true;
+  }
+}
 
-  // Start playing exactly on the next 4-beat boundary with looping.
-  newSource.loop = true;
-  newSource.start(time);
-};
+function makeDraggable(el, ondrag) {
+  el.addEventListener('dragstart', (event) => {
+    ondrag(event);
+    html.classList.add('dragging');
+  });
 
-function draw() {
-  // Calculate 4/4 beat position.
-  const time = context.currentTime - anchorTime;
-  const beat = (tempo / 60.0) * time;
-  const roundedBeat = 4.0 * Math.floor(beat / 4.0);
-  const wrappedBeat = beat - roundedBeat;
-  timeIndicator.value = wrappedBeat / 4;
+  el.addEventListener('dragend', (event) => {
+    html.classList.remove('dragging');
+  });
+}
 
-  const fullBeat = 1 + Math.trunc(beat / 4);
-  timeIndicatorText.textContent = `${fullBeat}.${1 + Math.trunc(wrappedBeat)}`;
+function makeDroppable(el, ondrop) {
+  el.ondragenter = (event) => {
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'copy';
+  };
+
+  el.ondragover = (event) => {
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'copy';
+  };
+
+  el.ondrop = (event) => {
+    event.preventDefault();
+    ondrop(event);
+  };
+}
+
+function play(context, decks) {
+  const now = context.currentTime;
+  anchorTime = now + 0.040;
 
-  requestAnimationFrame(draw);
+  decks[0].start(anchorTime);
+  decks[1].start(anchorTime);
+}
+
+function makeDial(containerEl, onchange) {
+  const controlEl = containerEl.querySelector('.control');
+  const textEl = containerEl.querySelector('.value');
+  const dial = new Nexus.Dial(controlEl, {mode: 'absolute'});
+  dial.on('change', (value) => {
+    textEl.innerText = onchange(value);
+  });
+  return dial;
 }
 
 async function init() {
   // Initialize audio
-  context = new AudioContext();
+  const context = new AudioContext();
 
   // Create post-compressor gain node.
   const postCompressorGain = context.createGain();
   postCompressorGain.gain.value = 1.4;
 
   postCompressorGain.connect(context.destination);
 
-  let compressor;
-
-  if (context.createDynamicsCompressor) {
-    // Create dynamics compressor to sweeten the overall mix.
-    compressor = context.createDynamicsCompressor();
-    compressor.connect(postCompressorGain);
-  } else {
-    // Compressor is not available on this implementation - bypass and simply
-    // point to destination.
-    compressor = postCompressorGain;
-  }
-
-  // Create sources
-  source1 = context.createBufferSource();
-  source2 = context.createBufferSource();
-
-  // Create gain nodes to control the volume for the two sources.
-  source1Gain = context.createGain();
-  source2Gain = context.createGain();
+  // Create dynamics compressor to sweeten the overall mix.
+  const compressor = context.createDynamicsCompressor();
+  compressor.connect(postCompressorGain);
 
   // Create pre-compressor gain node.
   const preCompressorGain = context.createGain();
```

Comment:
We should avoid the old factory pattern. We prefer:
```js
const preCompressorGain = new GainNode(context, {gain: 0.4});
```

Reply:
Also ditto for all other factory methods.

Reply:
Thanks for the review! Could you explain why this is preferred? The [spec](https://webaudio.github.io/web-audio-api/#factory-method) seems unopinionated and [MDN](https://developer.mozilla.org/en-US/docs/Web/API/GainNode/GainNode) strongly prefers the factory methods.

Reply:
I am not sure why MDN is documented that way, but it is wrong. We changed the spec because factory methods are generally bad, but especially for the extensibility. More info:
https://github.com/w3ctag/design-reviews/blob/main/2013/07/WebAudio.md#issue-constructibility--de-sugaring-of-static-methods

Also you can't use `GainNodeOptions` object with the factory method.

Reply:
Interesting! I'll follow up with MDN and change the code accordingly.

Reply:
Ah, I guess Safari compatibility might be a reason. Safari only supports the constructor since April 2021.

[new GainNode()](https://caniuse.com/mdn-api_gainnode_gainnode) has 84.4 % support
[createGain()](https://caniuse.com/mdn-api_baseaudiocontext_creategain) has 95.2 % support

Reply:
```
const context = new AudioContext();
```

I am quite certain the 15.6% of Safari won't pass the line above for the same reason. Have you tested the app on the most recent iOS?

Reply:
Fair point, the unprefixed AudioContext constructor has the same support in Safari. I'm fine with either changing the code back to increase compatibility, or changing it forward to show the modern API. Which SLO for browser compatibility do you suggest @hoch?

Reply:
As a side note: I test the web app on Safari 14.1.1 on macOS 11.4 and it works.

Reply:
Raised https://github.com/mdn/content/issues/7426 with MDN.

Reply:
Thanks! If its most recent version supports the modern Web Audio correctly, then we don't have any blocker! :)

Reply:
Done.